lib/ClangImporter/ClangClassTemplateNamePrinter.cpp

```diff
@@ -137,6 +137,17 @@ struct TemplateInstantiationNamePrinter
             Visit(type->getElementType().getTypePtr()) + ">")
         .str();
   }
+
+  std::string VisitArrayType(const clang::ArrayType *type) {
+    return (Twine("[") + Visit(type->getElementType().getTypePtr()) + "]")
+        .str();
+  }
+
+  std::string VisitConstantArrayType(const clang::ConstantArrayType *type) {
+    return (Twine("Vector<") + Visit(type->getElementType().getTypePtr()) +
```

Comment:
Where are these names coming from? Is there a documented convention we follow? The main reason for asking because I am wondering if this could potentially collide with a user-defined templated type named `Vector`.

Reply:
These names are made to look readable to a Swift developer, while preventing name collisions. The idea is that if/when we support instantiating C++ class templates from Swift without a typedef, these names would be very similar to the template signatures that Swift developers would use for C++ types. Currently the developer can never refer to the types using these names, so they are only visible in the module interfaces.

Here specifically there wouldn't be a collision, since for a user-defined `Vector<Element, 123>` type the generated Swift name would be `Vector<Element, _123_CUnsignedLong>`. This is more of quirk: this translation was implemented before value generics became a Swift feature, so the `_123_CUnsignedLong` was made to look like a type name to the parser. Eventually we would want to switch to something like `Vector<Element, 123 as CUnsignedLong>` for user-defined C++ type `Vector`. Note that the type of the literal (`CUnsignedLong`) still has to be a part of the generated name for disambiguation, so we'll never have a collision with `Vector<Element, 123>`

Reply:
Sounds fair, thanks!